packages/xo-server/src/api/vm.js

```diff
@@ -383,15 +383,25 @@ const delete_ = defer(async function(
   // Update resource sets
   let resourceSet
   if (
-    vm.type === 'VM' && // only regular VMs
+    (vm.type === 'VM' || vm.type === 'VM-snapshot') &&
     (resourceSet = xapi.xo.getData(vm._xapiId, 'resourceSet')) != null
   ) {
     await this.setVmResourceSet(vm._xapiId, null)::ignoreErrors()
     $defer.onFailure(() =>
-      this.setVmResourceSet(vm._xapiId, resourceSet)::ignoreErrors()
+      this.setVmResourceSet(vm._xapiId, resourceSet, true)::ignoreErrors()
     )
   }
 
+  await Promise.all(
+    vm.snapshots.map(async id => {
+      const { resourceSet } = this.getObject(id)
+      if (resourceSet !== undefined) {
+        await this.setVmResourceSet(id, null)::ignoreErrors()
+        $defer.onFailure(() => this.setVmResourceSet(id, resourceSet, true))
+      }
+    })
+  )
```

Comment:
Be careful with `Promise.all` and rejections, `$defer` does not work after the decorated function returned/threw.